server/src/main/java/org/opensearch/action/search/SearchRequestContext.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.lucene.search.TotalHits;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class holds request-level context for search queries at the coordinator node
+ *
+ * @opensearch.internal
+ */
+public class SearchRequestContext {
+    private long absoluteStartNanos;
+    private final Map<String, Long> phaseTookMap;
+    private TotalHits totalHits;
+    private Map<String, Integer> shardStats;
+
+    public SearchRequestContext() {
+        this.phaseTookMap = new HashMap<>();
+        this.totalHits = null;
+    }
+
+    protected void updatePhaseTookMap(String phaseName, Long tookTime) {
+        this.phaseTookMap.put(phaseName, tookTime);
+    }
+
+    protected Map<String, Long> phaseTookMap() {
+        return phaseTookMap;
+    }
+
+    protected void setAbsoluteStartNanos(long absoluteStartNanos) {
+        this.absoluteStartNanos = absoluteStartNanos;
+    }
+
+    protected long absoluteStartNanos() {
+        return absoluteStartNanos;
+    }
+
+    protected void setTotalHits(TotalHits totalHits) {
+        this.totalHits = totalHits;
+    }
+
+    protected TotalHits totalHits() {
+        return totalHits;
+    }
+
+    protected void setShardStats(Map<String, Integer> shardStats) {
+        this.shardStats = shardStats;
+    }
+
+    protected Map<String, Integer> shardStats() {
+        return shardStats;
+    }
+
+    public final String SEARCH_REQUEST_SLOWLOG_SHARD_TOTAL = "total";
+    public final String SEARCH_REQUEST_SLOWLOG_SHARD_SUCCESSFUL = "successful";
+    public final String SEARCH_REQUEST_SLOWLOG_SHARD_SKIPPED = "skipped";
+    public final String SEARCH_REQUEST_SLOWLOG_SHARD_FAILED = "failed";
```

Comment:
Instead of `String` keys, can we declare an `enum` and use an `EnumMap`?

Reply:
Right now we have `Map<String, Integer>` to store the shard stats `{total: 10, successful: 10, skipped: 0, failed: 0}`. If we use an `EnumMap<MyEnum, Integer>` instead, we lose the "String" mapping. Then we need to hard code "total", "successful", "skipped", "failed" in the output or maintain another `EnumMap<MyEnum, String>`. Am I understanding this correctly?

```
enum MyEnum {
    SEARCH_REQUEST_SLOWLOG_SHARD_TOTAL,
    SEARCH_REQUEST_SLOWLOG_SHARD_SUCCESSFUL,
    SEARCH_REQUEST_SLOWLOG_SHARD_SKIPPED,
    SEARCH_REQUEST_SLOWLOG_SHARD_FAILED
}
```

Reply:
The "String" mapping information can be encapsulated within Enum itself. You can check this out for reference: https://stackoverflow.com/questions/6667243/using-enum-values-as-string-literals

Reply:
That is a nice workaround. Done